docs/xks/developer-guide/best-practices.md

```diff
@@ -3,18 +3,125 @@ id: best-practices
 
-* [https://srcco.de/posts/web-service-on-kubernetes-production-checklist-2019.html](https://srcco.de/posts/web-service-on-kubernetes-production-checklist-2019.html)
+- **Run more than one replica in all environments**
+
+  Unless you have a special case you should run your application as more than one replica, even in non production environments. Apart from the more obvious fact that you get higher availability, it will help you catch bugs related to concurrency earlier. For example, if your application writes to a database and you want high availability in your production environment, add high availability to your non production environments as well. If you have a bug, e.g. some bad transactional logic in your application it is better to expose this bug in non production environments as well.
+
+- **Crash on unrecoverable errors**
+
+  Since it is easy to create new pods in Kubernetes it is better to just crash and start from a well defined state if your application encounters errors that are not recoverable.
+
+# Production Readiness checklist
+
+Please consider the following items before bringing your application to production. There exist a multitute of situations where items from this checklist does not make sense, but we strongly recommend that you at least consider every item in order to have as stable and well behaving applications as possible.
+
+## Read language specific docs
+
+Xenit provides some language specific documentation, e.g. [Xenit's Golang style guide](https://xenitab.github.io/docs/xenit-style-guide/golang) or [Xenit's Javascript/Typescript style guide](https://xenitab.github.io/docs/xenit-style-guide/javascript). These are summaries of experiences we have gathered while running production applications in XKF.
+
+## Readiness/Liveness probes
+
+Good probes are important if you want stable applications. They help with error reporting, but in our experience they are the most important when you deploy new versions.
+
+The short summary follows,
+
+1. Liveness probes are too powerful and in most situations you do not want them.
+2. You most likely want a basic readiness probe. This readiness probe should probably not involve other applications as this can lead to thundering herd problems. An endpoint that answers with **200** on **/healthz** on your http server for your application is a very good start.
+
+Your mileage may vary and only you can know what it means for your application to be ready to receive traffic. Consider reading this https://srcco.de/posts/kubernetes-liveness-probes-are-dangerous.html as it provides a more nuanced discussion when and where you want to use liveness/readiness probes.
+
+## Incoming HTTP
+
+TODO
+
+## Outgoing HTTP
+
+TODO
+
+## Database
+
+Your pod can be shut down by kubernetes. Make sure the following has been configured.
+
+1. Your application has graceful shutdown that closes connection to database.
```

Comment:
Mention something about sigterm and that you application needs to capture it and then close the connections.
We probably have some docs about this in some application specific part. Can link to that.